### PR TITLE
Rewrite 'react-native init' and 'react-native upgrade' without using Yeoman in preparation for templates support

### PR DESCRIPTION
**Pull request by mkonicek**
We'll be adding support for templates into 'react-native init', so you can create an app with tabbed navigation, Android drawer etc. to get started quickly.

In preparation for adding this feature, let's stop using Yeoman which we've wanted to do for quite a while - without Yeoman the code is shorter, easier to understand and we'll have less dependencies.

From user's perspective, this PR doesn't change the init experience - just replaces Yeoman with simpler code - see `copyProjectTemplateAndReplace`. The upgrade experience stays *almost* the same - we still ask how to resolve conflicts. The only thing missing is showing the difference between files but we'll replace that soon with https://github.com/ncuillery/rn-diff

**Test plan**

- Circle CI and Travis tests on this pull request (we run an e2e test both for iOS and Android: https://github.com/facebook/react-native/blob/master/scripts/run-ci-e2e-tests.js)
- `npm publish` to Sinopia as described here: https://github.com/facebook/react-native/blob/master/react-native-cli/README.md
- `react-native init AwesomeApp` creates an app, everything is named correctly, Java package is `com.awesomeapp`:

screenshot 2016-11-14 22 44 05

- `react-native run-ios` and `react-native run-android` - the app starts, Reload JS works (this is also tested by Circle CI and Travis):

screenshot 2016-11-15 16 28 13

screenshot 2016-11-14 23 27  #21

Enabled HMR, edited and saved index.android.js, JS was hot-reloaded:

screenshot 2016-11-16 15 51 43

- Made a change to .flowconfig, ran `react-native upgrade`, answered 'n' to the question to replace:

```
react-native upgrade
Upgrading project to react-native v1000.0.0
Check out the release notes and breaking changes: https://github.com/facebook/react-native/releases/tag/v1000.0.0
.flowconfig has changed in the new version.
Do you want to keep your .flowconfig or replace it with the latest version?
If you ever made any changes to this file, you'll probably want to keep it.
You can see the new version here: /Users/mkonicek/CLI_Rewrite_Test/node_modules/react-native/local-cli/templates/HelloWorld/.flowconfig
Do you want to replace .flowconfig? Answer y to replace, n to keep your version:
n
Keeping your .flowconfig
Successfully upgraded this project to react-native v1000.0.0

(Checked the file was kept intact)
```

- Made a change to .flowconfig, ran `react-native upgrade`, answered 'y' to the question to replace:

```
react-native upgrade
Upgrading project to react-native v1000.0.0
Check out the release notes and breaking changes: https://github.com/facebook/react-native/releases/tag/v1000.0.0
.flowconfig has changed in the new version.
Do you want to keep your .flowconfig or replace it with the latest version?
If you ever made any changes to this file, you'll probably want to keep it.
You can see the new version here: /Users/mkonicek/CLI_Rewrite_Test/node_modules/react-native/local-cli/templates/HelloWorld/.flowconfig
Do you want to replace .flowconfig? Answer y to replace, n to keep your version:
y
Replacing .flowconfig
Successfully upgraded this project to react-native v1000.0.0

(Checked the file was overwritten)
```

### CODE REVIEW

**Comment by bestander** [comment]
LGTM so far

**Comment by ericvicenti** [comment]
Looking pretty good.

I notice you forked the template/generated files. Please keep an eye on any changes to the old `generator` that you may need to sync over to your fork- I'd hate to see a regression because of that

**Comment by mkonicek** [comment]
@ericvicenti Thanks for looking! Totally agree. Just updated the template to the latest version. git recognizes the files have moved so we can see the changes in them.

File: local-cli/init/init.js [hunk 1]
```diff
@@ -8,48 +8,104 @@
...
 
-  env.register(
-    require.resolve(path.join(__dirname, '../generator')),
-    'react:app'
+  const args = Array.isArray(argsOrName)
+    ? argsOrName // ['AwesomeApp', '--verbose']
+    : [argsOrName].concat(process.argv.slice(4)); // 'AwesomeApp'
+
+  // args is ['AwesomeApp', '--verbose']
+  if (!args || args.lentgh == 0) {
```
Comment (bestander) [inline local-cli/init/init.js]: I assume there is no verbose here?
I think it is not needed, slows down npm even more

File: local-cli/init/init.js [hunk 2]
```diff
@@ -8,48 +8,104 @@
...
+    execSync(`yarn add react@${reactVersion}`);
+  } else {
+    console.log('Installing React...');
+    execSync(`npm install react@${reactVersion} --save --save-exact`);
+  }
+  if (!options['skip-jest']) {
+    const jestDeps = `jest babel-jest jest-react-native babel-preset-react-native react-test-renderer@${reactVersion}`;
+    if (yarnVersion) {
+      console.log('Adding Jest...');
+      execSync(`yarn add ${jestDeps} --dev --exact`);
```
Comment (bestander) [inline local-cli/init/init.js]: Martin, while you are at it, how about we provide a backdoor for CI systems and offline installs.
Recently Legocastle was not able to block a bug because we can't reproduce install offline.

Any ideas?
From the top of my head we could provide an override CLI command of some sort here.
And compose this command on CI to fake yarn install

File: package.json [hunk 3]
```diff
@@ -23,6 +23,7 @@
       "^[./a-zA-Z0-9$_-]+\\.png$": "RelativeImageStub"
     },
     "testPathIgnorePatterns": [
+      "local-cli/templates/",
       "/node_modules/",
```
Comment (bestander) [inline package.json]: Is there anymore yeoman left?
Can it be removed from package.json?

File: local-cli/generator/copyProjectTemplateAndReplace.js [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+    console.log(chalk.bold('new') + ' ' + relativeDestPath);
+    return 'overwrite';
+  } else if (contentChanged === 'changed') {
+    console.log(chalk.bold(relativeDestPath) + ' ' +
+      'has changed in the new version.\nDo you want to keep your ' + 
+      relativeDestPath + ' or replace it with the ' +
+      'latest version?\nIf you ever made any changes ' +
+      'to this file, you\'ll probably want to keep it.\n' +
+      'You can see the new version here: ' + absoluteSrcFilePath + '\n' +
+      'Please answer "keep" / "replace" for ' + relativeDestPath + ':');
```
Comment (ncuillery) [inline local-cli/generator/copyProjectTemplateAndReplace.js]: Maybe we could have a more "yeoman-ish" prompt with the status clearly displayed in colors and a Yes/No answer (default Yes).
The log about the changes and the HelloWorld app location should be displayed once and for all at the beginning.

Typing "keep" or "replace" for each file (typically 5 to 10 times) is boring.

File: local-cli/generator/copyProjectTemplateAndReplace.js [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+    const answer = prompt();
+    if (answer == 'replace') {
```
Comment (ncuillery) [inline local-cli/generator/copyProjectTemplateAndReplace.js]: Double equal :smile:

Comment (ncuillery) [inline local-cli/generator/copyProjectTemplateAndReplace.js]: If the user makes a typo in "replace", the file is kept. It could lead to a lot of madness :p

File: local-cli/generator/copyProjectTemplateAndReplace.js [hunk 6]
```diff
@@ -0,0 +1,95 @@
...
+      console.log('Replacing ' + relativeDestPath);
+      return 'overwrite';
+    } else {
+      console.log('Keeping your ' + relativeDestPath);
+      return 'keep';
+    }
+  } else if (contentChanged === 'identical') {
+    return 'keep';
+  } else {
+    throw new Error(`Unkown file changed state: {relativeDestPath}, {contentChanged}`);
```
Comment (ncuillery) [inline local-cli/generator/copyProjectTemplateAndReplace.js]: Incorrect template string placeholders: missing `$`

File: local-cli/generator/copyProjectTemplateAndReplace.js [hunk 7]
```diff
@@ -0,0 +1,95 @@
+/**
+ * Copyright (c) 2015-present, Facebook, Inc.
+ * All rights reserved.
+ *
+ * This source code is licensed under the BSD-style license found in the
+ * LICENSE file in the root directory of this source tree. An additional grant
+ * of patent rights can be found in the PATENTS file in the same directory.
+ *
+ * @flow
```
Comment (mkonicek) [inline local-cli/generator/copyProjectTemplateAndReplace.js]: It's not good catch!

File: local-cli/generator/copyProjectTemplateAndReplace.js [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+  if (!newProjectName) throw new Error('Need a project name');
+
+  walk(srcPath).forEach(absoluteSrcFilePath => {
+
+    // 'react-native upgrade'
+    if (options && options.upgrade) {
+      // Don't upgrade these files
+      const fileName = path.basename(absoluteSrcFilePath);
+      // TODO __tests__/index.ios.js
+      // TODO __tests__/index.android.js
```
Comment (mkonicek) [inline local-cli/generator/copyProjectTemplateAndReplace.js]: Will fix this, thanks for the reminder!

File: local-cli/init/init.js (hunk 1 above)
Comment (mkonicek) [inline local-cli/init/init.js]: The args is for example ['AwesomeApp', '--verbose'].

Do you mean supporting `--verbose` is not needed? It was added by the community to debug when 'npm install' helps, I'd keep it.

File: local-cli/init/init.js (hunk 2 above)
Comment (mkonicek) [inline local-cli/init/init.js]: Interesting, let's talk about so I understand how it should work. 

I think this PR is getting very large - can it be done as a separate commit?

File: local-cli/generator/copyProjectTemplateAndReplace.js (hunk 4 above)
Comment (mkonicek) [inline local-cli/generator/copyProjectTemplateAndReplace.js]: Good point that typing "keep" or "replace" is boring. Can make it "do you want to overwrite the file? y/n"?

File: local-cli/generator/copyProjectTemplateAndReplace.js (hunk 5 above)
Comment (mkonicek) [inline local-cli/generator/copyProjectTemplateAndReplace.js]: Good catch about `==`, thanks!

If people make a mistake and keep files they wanted to overwrite, they can run 'react-native upgrade' again to start over.

File: local-cli/generator/copyProjectTemplateAndReplace.js (hunk 6 above)
Comment (mkonicek) [inline local-cli/generator/copyProjectTemplateAndReplace.js]: Awesome, thank you!

File: local-cli/generator/copyProjectTemplateAndReplace.js (hunk 4 above)
Comment (mkonicek) [inline local-cli/generator/copyProjectTemplateAndReplace.js]: I prefer to default to 'n' - overwriting the files throws away the changes people made in them.

File: package.json (hunk 3 above)
Comment (mkonicek) [inline package.json]: Plan remove it in a separate celebratory commit :)

**Comment by mkonicek** [comment]
TODO [e2e test](https://circleci.com/gh/facebook/react-native/13243?utm_campaign=vcs-integration-link&utm_medium=referral&utm_source=github-build-link) fails with:

    Wrong version of Flow. The config specifies version ^0.34.0 but this is version 0.35.0
    Flow check does not pass

**Comment by bestander** [comment]
it was updated just a day ago

**Comment by mkonicek** [comment]
Flow check should now be fixed, Dan Caspi has a fix for the iOS compile error (not landed yet).

**Comment by hramos** [comment]
@facebook-github-bot shipit

**Comment by mkonicek** [comment]
Closed in https://github.com/facebook/react-native/commit/a477aec10d29b4651944cb4292daf06dfa953ea3.
